server/src/metadata.rs

```diff
@@ -0,0 +1,1310 @@
+        initialize_model_table_with_tag_values(&mut metadata_manager, &["tag_value1"]);
+
+        // Lookup the keys for the fallback column by only requesting tag columns.
+        let keys = metadata_manager
+            .compute_keys_using_fields_and_tags("model_table", &Some(vec![0]), 10, &[])
+            .unwrap();
+        assert_eq!(1, keys.len());
+    }
+
+    #[test]
+    fn test_compute_the_keys_for_a_specific_field_column_for_model_table() {
+        // Save a model table to the metadata database.
+        let temp_dir = tempfile::tempdir().unwrap();
+        let mut metadata_manager = test_util::get_test_metadata_manager(temp_dir.path());
+        initialize_model_table_with_tag_values(
+            &mut metadata_manager,
+            &["tag_value1", "tag_value2"],
+        );
+
+        // Lookup all keys for the table by not requesting keys for columns.
+        let keys = metadata_manager
+            .compute_keys_using_fields_and_tags("model_table", &None, 10, &[])
+            .unwrap();
+        assert_eq!(4, keys.len());
+
+        // Lookup keys for the table by requesting keys for a specific field column.
+        let keys = metadata_manager
+            .compute_keys_using_fields_and_tags("model_table", &Some(vec![1]), 10, &[])
+            .unwrap();
+        assert_eq!(2, keys.len());
+    }
+
+    #[test]
+    fn test_compute_the_keys_for_a_specific_tag_value_for_model_table() {
+        // Save a model table to the metadata database.
+        let temp_dir = tempfile::tempdir().unwrap();
+        let mut metadata_manager = test_util::get_test_metadata_manager(temp_dir.path());
+        initialize_model_table_with_tag_values(
+            &mut metadata_manager,
+            &["tag_value1", "tag_value2"],
+        );
+
+        // Lookup all keys for the table by not requesting keys for a tag value.
+        let keys = metadata_manager
+            .compute_keys_using_fields_and_tags("model_table", &None, 10, &[])
+            .unwrap();
+        assert_eq!(4, keys.len());
+
+        // Lookup keys for the table by requesting keys for a specific tag value.
+        let keys = metadata_manager
+            .compute_keys_using_fields_and_tags("model_table", &None, 10, &[("tag", "tag_value1")])
+            .unwrap();
+        assert_eq!(2, keys.len());
+    }
+
+    fn initialize_model_table_with_tag_values(
+        metadata_manager: &mut MetadataManager,
+        tag_values: &[&str],
+    ) {
+        let model_table_metadata = test_util::get_model_table_metadata();
+        metadata_manager
+            .save_model_table_metadata(&model_table_metadata, vec![1, 2, 4, 8])
+            .unwrap();
+
+        for tag_value in tag_values {
+            let tag_value = tag_value.to_string();
+            metadata_manager
+                .get_or_compute_tag_hash(&model_table_metadata, &vec![tag_value])
+                .unwrap();
+        }
+    }
+
+    #[test]
+    fn create_metadata_database_tables() {
```

Comment:
This test function is missing the `test_` prefix.

Reply:
Fixed.